src/AppInstallerCommonCore/Settings.cpp

```diff
@@ -128,66 +111,143 @@ namespace AppInstaller::Settings
+            bool Set(std::string_view value) override
             {
-                std::filesystem::path settingFileName = GetPath(name, true);
+                EnsureParentPath();
 
-                std::ofstream stream(settingFileName, std::ios_base::out | std::ios_base::binary | std::ios_base::trunc);
+                std::ofstream stream(m_settingFile, std::ios_base::out | std::ios_base::binary | std::ios_base::trunc);
                 THROW_LAST_ERROR_IF(stream.fail());
                 stream << value << std::flush;
                 THROW_LAST_ERROR_IF(stream.fail());
+
+                return true;
             }
 
-            void Remove(const std::filesystem::path& name) override
+            void Remove() override
             {
-                std::filesystem::path settingFileName = GetPath(name);
-
-                std::filesystem::remove(settingFileName);
+                std::filesystem::remove(m_settingFile);
             }
 
-            std::filesystem::path PathTo(const std::filesystem::path& name) override
+            std::filesystem::path PathTo() override
             {
-                return GetPath(name);
+                return m_settingFile;
             }
 
         private:
-            std::filesystem::path GetPath(const std::filesystem::path& name, bool createParent = false)
+            void EnsureParentPath()
             {
-                std::filesystem::path result = m_root;
+                std::filesystem::create_directories(m_settingFile.parent_path());
+            }
 
-                if (name.has_parent_path())
+            std::filesystem::path m_settingFile;
+        };
+
+        // A settings container that manages safely writing to its value with exchange semantics.
+        // Only allows Set to succeed if the hash value of the setting is the same as the last time it was read.
+        struct ExchangeSettingsContainer : public details::ISettingsContainer
+        {
+            ExchangeSettingsContainer(std::unique_ptr<ISettingsContainer>&& container, const std::string_view& name) :
+                m_container(std::move(container)), m_name(name) {}
+
+            std::unique_ptr<std::istream> Get() override
+            {
+                return GetInternal(m_hash);
+            }
+
+            bool Set(std::string_view value) override
+            {
+                THROW_HR_IF(E_UNEXPECTED, value.size() > std::numeric_limits<uint32_t>::max());
+
+                // If Set is called without ever reading the value, then we can assume that caller wants
+                // to overwrite it regardless. Also, we don't have any previous value to compare against
+                // anyway so the only other option would be to always reject it.
+                if (m_hash)
                 {
-                    result /= name.parent_path();
+                    std::optional<SHA256::HashBuffer> currentHash;
+                    std::ignore = GetInternal(currentHash);

```

Comment:
>GetInternal(currentHash) [](http://example.com/codeflow?start=34&length=24)

I thought we don't like out parameter? Maybe it's ok for internal implementation

Reply:
It has its places, and I think that in this internal case it works well to enable the side effect to be controlled without adding type overhead.